src/pipeline.rs

```diff
@@ -0,0 +1,192 @@
+#[serde(rename_all = "snake_case")]
+pub enum CheckConf {
+    SpecialValueCheck(SpecialValueCheckConf),
+    RangeCheck(RangeCheckConf),
+    RangeCheckDynamic(RangeCheckDynamicConf),
+    StepCheck(StepCheckConf),
+    SpikeCheck(SpikeCheckConf),
+    FlatlineCheck(FlatlineCheckConf),
+    BuddyCheck(BuddyCheckConf),
+    Sct(SctConf),
+    ModelConsistencyCheck(ModelConsistencyCheckConf),
+    #[serde(skip)]
+    Dummy,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct SpecialValueCheckConf {
+    pub special_values: Vec<f32>,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct RangeCheckConf {
+    pub max: f32,
+    pub min: f32,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct RangeCheckDynamicConf {
+    pub source: String,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct StepCheckConf {
+    pub max: f32,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct SpikeCheckConf {
+    pub max: f32,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct FlatlineCheckConf {
+    pub max: u8,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct BuddyCheckConf {
+    pub radii: Vec<f32>,
+    pub nums_min: Vec<u32>,
+    pub threshold: f32,
+    pub max_elev_diff: f32,
+    pub elev_gradient: f32,
+    pub min_std: f32,
+    pub num_iterations: u32,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct SctConf {
+    pub num_min: usize,
+    pub num_max: usize,
+    pub inner_radius: f32,
+    pub outer_radius: f32,
+    pub num_iterations: u32,
+    pub num_min_prof: usize,
+    pub min_elev_diff: f32,
+    pub min_horizontal_scale: f32,
+    pub vertical_scale: f32,
+    pub pos: Vec<f32>,
+    pub neg: Vec<f32>,
+    pub eps2: Vec<f32>,
+    pub obs_to_check: Option<Vec<bool>>,
+}
+
+#[derive(Debug, Deserialize, PartialEq, Clone)]
+pub struct ModelConsistencyCheckConf {
+    pub model_source: String,
+    pub model_args: String,
+    pub threshold: f32,
+}
+
+#[derive(Error, Debug)]
+pub enum Error {
+    /// Generic IO error
+    #[error("io error: {0}")]
+    Io(#[from] std::io::Error),
+    /// TOML deserialize error
+    #[error("failed to deserialize toml: {0}")]
+    TomlDeserialize(#[from] toml::de::Error),
+    /// The directory contained something that wasn't a file
+    #[error("the directory contained something that wasn't a file")]
+    DirectoryStructure,
+    /// Pipeline filename could not be parsed as a unicode string
+    #[error("pipeline filename could not be parsed as a unicode string")]
+    InvalidFilename,
+}
+
+/// Given a pipeline, derive the number of leading and trailing points per timeseries needed in
+/// a dataset, for all the intended data to be QCed by the pipeline
+pub fn derive_num_leading_trailing(pipeline: &Pipeline) -> (u8, u8) {
+    // TODO: would there be somewhere better for this match to live?
+    // perhaps harness or even olympian, but annoyingly it depends on things in this module
```

Comment:
I think it makes sense here, we are populating the Pipeline struct with data concerning the pipeline.

Reply:
Ah, my bad, I forgot about the match we have in `harness::run_test`! Maybe we can extract the closure we have in the map inside `derive_num_leading_trailing` to a `CheckConf` method? Then in `run_test` we can call said method after declaring `step_name` 
```rust
let (num_leading_per_run, num_trailing_per_run) = step.check.get_num_extra_points();
``` 
I don't know if that makes things better though...

Reply:
Ahh no it can't be const because the limit for flatline check (previously "freeze check") comes from the config and isn't known at compile time. 

BUT I realized we can just define these constants in `mod harness` (or better yet olympian when we get around to it there), then import and use them in `mod pipeline`, so I'm going to do that.

Reply:
Done